crates/arroyo-operator/src/context.rs

```diff
@@ -80,6 +84,128 @@ impl WatermarkHolder {
     }
 }
 
+/// A wrapper for an UnboundedSender<QueueItem> that bounds by the number of rows within
+/// a batch rather than the number of batches
+#[derive(Clone)]
+pub struct BatchSender {
+    size: u32,
+    tx: UnboundedSender<QueueItem>,
+    queued_messages: Arc<AtomicU32>,
```

Comment:
What about using tokio's Semaphores rather than Atomics? I think that'll let us automatically integrate with async.

Reply:
The semaphore API didn't seem suitable for this unfortunately. The acquire call returns a permit which—when dropped—releases the slots. However, for this we need to acquire the permit in the producer and release it in the consumer.